opendevin/memory/condenser.py

```diff
@@ -1,26 +1,199 @@
+from typing import Callable
+
 from opendevin.core.logger import opendevin_logger as logger
+from opendevin.core.utils import json
 from opendevin.llm.llm import LLM
 
+MAX_TOKEN_COUNT_PADDING = 1024
+
 
 class MemoryCondenser:
-    def condense(self, summarize_prompt: str, llm: LLM):
+    """
+    Condenses the prompt with a call to the LLM.
+    """
+
+    def __init__(
+        self,
+        action_prompt: Callable[..., str],
+        summarize_prompt: Callable[[list[dict], list[dict]], str],
+    ):
+        """
+        Initialize the MemoryCondenser with the action and summarize prompts.
+
+        action_prompt is a callable that returns the prompt that is about to be sent to the LLM.
+        The prompt callable will be called with default events and recent events as arguments.
+        summarize_prompt is a callable that returns a specific prompt that tells the LLM to summarize the recent events.
+        The prompt callable will be called with default events and recent events as arguments.
+
+        Parameters:
+        - action_prompt: The function to generate an action prompt. The function should accept default events and recent events as arguments.
+        - summarize_prompt: The function to generate a summarize prompt. The function should accept default events and recent events as arguments.
+        """
+        self.action_prompt = action_prompt
+        self.summarize_prompt = summarize_prompt
+
+    def condense(
+        self,
+        llm: LLM,
+        default_events: list[dict],
+        recent_events: list[dict],
+        background_commands: list | None = None,
+    ) -> tuple[list[dict], bool]:
         """
-        Attempts to condense the monologue by using the llm
+        Attempts to condense the recent events of the monologue by using the llm, if necessary. Returns the condensed recent events if successful, or False if not.
+
+        It includes default events in the prompt for context, but does not alter them.
+        Condenses the monologue using a summary prompt.
+        Checks if the action_prompt (including events) needs condensation based on token count, and doesn't attempt condensing if not.
+        Returns unmodified list of recent events if it is already short enough.
 
         Parameters:
-        - llm (LLM): llm to be used for summarization
+        - llm: LLM to be used for summarization.
+        - default_events: List of default events that should remain unchanged.
```

Comment:
I see, I'm looking at CodeAct, and these `default_events` are kind of what CodeAct has as `EXAMPLES`, and maybe similar to `GITHUB_MESSAGE` and `COMMAND_DOCS`. For CodeAct these are included in system prompt and that's it.

There is one difference though, for Monologue, they are _Events_:
- it has two lists of "first lessons"
- one is parsed as actual events `default_events`("initial thoughts"= hey it's waking up in this docker world 😄)
- the other is documentation of actions it knows
- they both are supposed to be in the "prefix" of sorts, that is, in the action prompt at all times.

_Maybe_ this particular Monologue list is not even needed in the summarize prompt, only action prompt... I thought it makes sense to give the LLM context with examples here too, but I'm not too sure.

On the other hand, I was also thinking that it's possible that we add others, not hard-coded in the prompts, but obtained at runtime from another place ("action library", "skills" library). That may look like the GITHUB_MESSAGE, with or without an example. I thought that for monologue, that means it could also add some actions/observations to its "real-like" `default_events` list.